### PR TITLE
Add drives subsites and sync user permissions

### PR DESCRIPTION
**Pull request by praveen-elastic**
Following changes have been made as a part of this PR:

- Support for drive items (files and folders)
- Support for Sub sites (sites inside a site)
- Implementation for fetching user-level permissions as an independent process
- Modified approach for multiprocessing
- Implementation for providing a single daemon process for all different functions in the connector

### CODE REVIEW

**Review by wangch079 — COMMENTED** [review]
Partially reviewed.

File: fetch_index.py [hunk 1]
```diff
@@ -27,7 +27,8 @@
 ITEM = "item"
 SITES = "sites"
 LISTS = "lists"
-ITEMS = "items"
+ITEMS = "list_items"
+DRIVES = "drive_items"
```
Comment (wangch079) [inline fetch_index.py:31]: `ITEMS` is a bit confusing. Maybe considering rename to `LIST_ITEMS` and `DRIVE_ITEMS`?

File: sync_user_permissions.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+                        user_name, exception
+                    )
+                )
+
+    def sync_permissions(self):
+        """ This method when invoked, checks the permission of SharePoint users and update those user
+            permissions in the Workplace Search.
+        """
+        rows = {}
+        if (os.path.exists(self.mapping_sheet_path) and os.path.getsize(self.mapping_sheet_path) > 0):
```
Comment (wangch079) [inline sync_user_permissions.py:92]: How is this `Sharepoint user` to `Workplace user` mapping managed?

File: fetch_index.py [hunk 3]
```diff
@@ -298,52 +312,136 @@ def index_items(self, lists, ids, index):
...
+                DRIVES,
+            )
+            if not response_data:
+                logger.info("No item was created for the library %s in this interval: start time: %s and end time: %s" % (value[1], self.start_time, self.end_time))
+                continue
+            logger.info(
+                "Successfully fetched and parsed %s drive item response for library: %s from SharePoint"
+                % (len(response_data), value[1])
+            )
+            schema_drive = self.get_schema_fields(DRIVES)
```
[the remark this answers is not in the document]
Comment (praveen-elastic) [reply fetch_index.py:406]: Makes sense. Updating it.

File: sync_user_permissions.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        or puts the connector to sleep
+    """
+    logger.info("Starting the permission indexing..")
+    config = Configuration("sharepoint_connector_config.yml", logger)
+    data = config.configurations
+
+    while True:
+        enable_permission = data.get("enable_document_permission")
+        permission_indexer = SyncUserPermission(data)
+        if not enable_permission:
```
[the remark this answers is not in the document]
Comment (praveen-elastic) [reply sync_user_permissions.py:122]: Got it. I'll implement that.

File: sync_user_permissions.py (hunk 2 above)
Comment (praveen-elastic) [reply sync_user_permissions.py:92]: Hey, SharePoint user to Workplace user mapping is done using a csv file whose path is entered by the user in the config.yml file. The first entry in each row is the username in SharePoint and the second entry is the username in Workplace.

File: fetch_index.py (hunk 1 above)
Comment (praveen-elastic) [reply fetch_index.py:31]: Correct. It should be `LIST_ITEMS`. Updating this

File: deindex.py [hunk 5]
```diff
@@ -65,7 +65,7 @@ def deindexing_items(self, collection, ids):
             for site_url in delete_site:
                 global_ids_items.pop(site_url)
         else:
-            logger.info("No list-items found to be deleted for collection: %s" % collection)
+            logger.info("No list-items or drive-items found to be deleted for collection: %s" % collection)
```
[the remark this answers is not in the document]
Comment (praveen-elastic) [reply deindex.py:68]: Hey @wangch079, thanks for reviewing the PR. I'll update this.

File: sync_user_permissions.py (hunk 2 above)
Comment (wangch079) [reply sync_user_permissions.py:100]: So it's manually maintained?

File: usergroup_permissions.py [hunk 6]
```diff
@@ -1,8 +1,8 @@
-from sharepoint_utils import encode
 from elastic_enterprise_search import WorkplaceSearch
 SITES = "sites"
 LISTS = "lists"
-ITEMS = "items"
+ITEMS = "list_items"
+DRIVES = "drive_items"
```
Comment (wangch079) [inline usergroup_permissions.py:5]: Maybe update here as well?

File: sync_user_permissions.py (hunk 2 above)
Comment (praveen-elastic) [reply sync_user_permissions.py:100]: Yes, the user will manually fill in the mappings csv file which would have SharePoint to workplace user mapping. It is a part of configuration.

File: usergroup_permissions.py (hunk 6 above)
Comment (praveen-elastic) [reply usergroup_permissions.py:5]: Sure, updating this.

File: sync_user_permissions.py (hunk 2 above)
Comment (nirav-elastic) [reply sync_user_permissions.py:100]: Yes @wangch079 , so the user will have to fill in the mappings.csv file as a part of the configuration step which would have Sharepoint->Workplace user mappings before they start the connector. The connector will then automatically iterate over that CSV mappings and translate the users

**Review by wangch079 — APPROVED** [review]

**Pull request merged**